### ISSUE TITLE
How to invalidate provider after widget disposed

### ISSUE THREAD

**Issue by claudius-kienle** [rrousselGit/riverpod#1704]
I currently use riverpod as state management all over my app including a lot of widges that `watch` the providers' state.
I now want to invalidate all providers upon user logout to clear any user-related data stored in the providers but struggeling how to implement this.

**Approach I tried**
I tried multiple possiblities to invalidate the providers, but all of which do not work properly and causing the providers not just to invalidate but also directly refresh:
- If I invalidate the providers directly after the logout was started, all "logged-in" widgets still watch the providers' state causing them to refresh.
- If I invalidate the providers after the login page finished loading (via. `WidgetsBinding.instance.addPostFrameCallback`), the "logged-in" widgets still weren't disposed and again causing the providers to refresh. I use GoRouter for Navigation.
- Invalidating the providers in the widget's `dispose()` function does not work since the *widget's element tree is no longer stable*.

**Expected behavior**
I need some functionality to first wait until all widgets listening to privoders are disposed before invalidating the providers' states. Currently, I am able to do this with a `Future.delayed` which just triggers the invalidation after 2 seconds. However, I do not think this is a gread solution.

Is there already some solution to this scenario?

**Comment by rrousselGit** [comment]
Why not use autoDispose?

**Comment by claudius-kienle** [comment]
all Providers execute API calls when they're created. To minimize API calls and also loading times, I want to retain the data when the user leaves the page and again visits it. I only want to invalidate the Providers once the user logs out.

**Comment by rrousselGit** [comment]
This should be naturally taken care of by `ref.watch`. I'd expect providers to "ref.watch" the provider that contains user info. Then on logout, they'd recompute by themselves

Alternatively you may be looking for https://github.com/rrousselGit/riverpod/issues/1664

**Comment by claudius-kienle** [comment]
No sorry #1664 is not what I'm looking for. 
If i try to invalidate a `StateNotifierProvider` after logout, it again directly recreates itself (like calling the `refresh` function). I don't want the provider to recreate itself as this always executes an API call, but the user is already logged out.

I now figured out that all `StateNotifierProvider` get refreshed because of
```dart
void _performRefresh() {
    /// No need to traverse entries from top to bottom, because refreshing a
    /// child will automatically refresh its parent when it will try to read it
    for (var i = 0; i < _stateToRefresh.length; i++) {
      final element = _stateToRefresh[i];
      if (element.hasListeners) element.flush();
    }
  }
```
since `element.hasListeners` evalutates to `true`. 
This is the case because the element `_NotifierProviderElement<T, Y>` has one entry as `dependents` being `ProviderElement<Y>`.
Consequently, `dependents` is not empty and `hasListeners` `true`. 
Sadly, I'm not deep enough in the riverpod code base to understand those to classes and why `dependents` is not empty. Any suggestions?

**Comment by rrousselGit** [comment]
> No sorry https://github.com/rrousselGit/riverpod/issues/1664 is not what I'm looking for.

It should be. You're not using autoDispose because you want to preserve the state if the user leaves then go back to a screen rapidly.

That's what cacheTime is about. You're supposed to use autoDispose, then set some cacheTime. For that duration, the request will be cached.

> Sadly, I'm not deep enough in the riverpod code base to understand those to classes and why dependents is not empty. Any suggestions?

It means your provider is still listened by something, whatever that is.

**Comment by claudius-kienle** [comment]
Thanks for your quick replys! I really apprechiate your help:)

> It should be. You're not using autoDispose because you want to preserve the state if the user leaves then go back to a screen rapidly.
But I would like to retain the state regardless if the user is going back to the back within a few seconds or a couple of minutes. Therefore, this approach wouldn't help me. Especially if the user logs out and then in with a different account within the cache time, he/she would see data from a different user account.

> It means your provider is still listened by something, whatever that is.
I now investigated the problem a bit more and am able to reproduce the issue successfully in [this](https://github.com/claudius-kienle/riverpod-provider-bug.git) repo.
If I open the Details Page a normal Provider and a StateNotifierProvider get initialized (note the `initialized` messages in the logs). Closing the details page and then clicking on `Clear Providers` invalidates both providers with `ref.invalidate(providers).
Since nothing listens on the providers they should only invalidate and not refresh, which is also what the `Provider` does (no `initialized` message). On the other hand, the `StateNotifierProvider` directly inialized itself again (`inialized` message popping up), even though there is no one listening.

I think, the StateNotifierProvider should not refresh itself since no one is listening, or is this behavior intended?

**Comment by rrousselGit** [comment]
That's the thing: If your provider is refreshing, it _is_ listened. 

I would challenge the fact that nothing listens to your provider. Something definitely is.

**Comment by claudius-kienle** [comment]
Could you please look at the code sample I provided and tell me where a component still listens to the StateNotifier? I don't think there is anyone listening

**Comment by claudius-kienle** [comment]
I closed this issue in favor of #1708
